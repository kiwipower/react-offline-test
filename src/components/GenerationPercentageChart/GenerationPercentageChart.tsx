import React, { 
    useState, useEffect, 
    Fragment, useContext 
} from 'react'

import config from '../../config';

import DateSpanViewer from "../DateSpanViewer";
import GenerationMixItem from "../GenerationMixItem";
import Loading from "../shared/Loading";
import GenerationDataContext from '../../store/GenerationData/Context';
import { getGenerationData } from '../../store/GenerationData/Actions';

export default function GenerationPercentageChart() {

    const [loading, setLoading] = useState(false);
    
    const generationDataContext = useContext(GenerationDataContext);

    useEffect(() => {
        loadData();
    }, [])

    const loadData = async () => {
        if (loading) {
            return;
        }
        setLoading(true);
        generationDataContext.dispatch(await getGenerationData());
        setLoading(false);
    }

    const refreshData = (e: React.MouseEvent<HTMLAnchorElement, MouseEvent>) => {
        e.preventDefault();
        loadData();
    };

    const renderContent = () => {
        if (loading) {
            return (<Loading />);
        }
    
        if (generationDataContext.state.lastError) {
            return (
                <div className="alert alert-danger">
                    {generationDataContext.state.lastError.message}
                </div>
            );
        }

        const { state: { currentData } } = generationDataContext;

        if (currentData === null) {
            return null;
        }

        return (
            <Fragment>
                <div className=" card date-container">
                    <DateSpanViewer 
                        from={currentData.from}
                        to={currentData.to} />
                </div>
                <div className="bar-chart">
                    {currentData.generationmix.map((gdata, i) => {
                        const key = `data-${i}`;
                        const color = config.barColors[i];
                        return (<GenerationMixItem 
                                    key={key}
                                    gdata={gdata} 
                                    color={color} />)
                    })}
                </div>
            </Fragment>
        );
    }

    return (
        <div>
            <div className="row my-4 justify-content-center">
                <a className={'btn btn-outline-primary ' + loading || ' btn-disabled'} 
                    id="refresh-data-button"
                    href="" onClick={refreshData}>
                    {loading ? 'Loading...' : 'Refresh Data'} 
                </a>
            </div>
            {renderContent()}
        </div>
    )
}
