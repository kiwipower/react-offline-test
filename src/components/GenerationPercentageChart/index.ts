import GenerationPercentageChart from "./GenerationPercentageChart";

export default GenerationPercentageChart;