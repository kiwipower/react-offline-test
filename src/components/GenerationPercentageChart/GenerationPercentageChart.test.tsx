import React from 'react';
import { shallow } from 'enzyme';

import GenerationPercentageChart from './GenerationPercentageChart';

describe('<GenerationPercentageChart />', () => {

    const renderComponent = () => (<GenerationPercentageChart />);

    it('renders without error', async () => {
        const component = shallow(renderComponent());
        expect(component).toMatchSnapshot();
    });

});