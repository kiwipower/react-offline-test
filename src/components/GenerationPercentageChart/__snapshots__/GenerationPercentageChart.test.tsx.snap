// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GenerationPercentageChart /> renders without error 1`] = `
<div>
  <div
    className="row my-4 justify-content-center"
  >
    <a
      className="btn btn-outline-primary false"
      href=""
      id="refresh-data-button"
      onClick={[Function]}
    >
      Refresh Data
    </a>
  </div>
</div>
`;
